include/rx/ranges.hpp

```diff
@@ -1998,6 +1972,114 @@ struct reverse {
     }
 };
 
+/// Return a range flattening one level of nesting in a range of ranges.
+///
+/// Supply a template parameter bigger than 1 to flatten multiple levels at once.
+template <size_t Depth>
+struct flatten;
+
+template <>
+struct flatten <0> {
+    template <class InputRange>
+    [[nodiscard]] constexpr auto operator()(InputRange&& input) const noexcept {
+        return std::forward<InputRange>(input);
+    }
+};
+
+template <>
+struct flatten <1> {
+    template <class R>
+    struct Range {
+        using S = get_range_type_t<get_output_type_of_t<R>>;
+        using output_type = get_output_type_of_t<S>;
+
+        static constexpr bool is_finite = is_finite_v<R> && is_finite_v<S>;
+        static constexpr bool is_idempotent = is_idempotent_v<S>;
+
+        R outer_range;
+        RX_OPTIONAL<S> inner_range;
+
+        template <class Rx>
+        constexpr explicit Range(Rx input) : outer_range(std::forward<Rx>(input)) {
+            while (RX_LIKELY(!outer_range.at_end())) {
+                inner_range.emplace(as_input_range(outer_range.get()));
+                if (RX_LIKELY(!inner_range->at_end())) {
+                    return;
+                }
+            }
+            inner_range.reset();
+        }
+
+        [[nodiscard]] constexpr bool at_end() const noexcept {
+            return !bool(inner_range);
+        }
+
+        [[nodiscard]] constexpr output_type get() const noexcept {
+            RX_ASSERT(!at_end());
+            return inner_range->get();
+        }
+
+        constexpr void next() noexcept {
+            RX_ASSERT(!at_end());
+
+            inner_range->next();
+            if (RX_LIKELY(!inner_range->at_end())) {
+                return;
+            }
+
+            for (;;) {
+                outer_range.next();
+                if (RX_UNLIKELY(outer_range.at_end())) {
+                    break;
+                }
+
+                inner_range.emplace(as_input_range(outer_range.get()));
+                if (RX_LIKELY(!inner_range->at_end())) {
+                    return;
+                }
+            }
+            inner_range.reset();
+        }
+
+        [[nodiscard]] constexpr size_t size_hint() const noexcept {
+            if (at_end()) {
+                return 0;
+            }
+
+            auto r = outer_range.size_hint();
+            auto s = inner_range->size_hint();
+            auto rs = r * s;
+            if (!r) {
+                return s;
+            } else if (!s) {
+                return r;
+            } else if (rs < r || rs < s) {
```

Comment:
`r == std::numeric_limits<size_t>::max() || s == std::numeric_limits<size_t>::max()` would be clearer IMO, you could also move the calcul of `rs` in the final else branch.

Reply:
You can use `R::is_finite` and `S::is_finite` with `if constexpr` to statically return `numeric_limits<size_t>::max()`.

Note that the contract for `size_hint()` was changed in #10 to return `numeric_limits<size_t>::max()` when the output length is unknown. This was done to keep the possibility of sometimes treating the hint as an exact value when the output length is actually precisely known, which in turn would allow specializations of combinators that can be more easily auto-vectorized.